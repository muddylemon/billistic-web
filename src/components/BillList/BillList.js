import React from "react"
import "./bill_list.scss"
import { withState, compose } from "recompose"

import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'
import Checkbox from '@material-ui/core/Checkbox'


const handleMarkPaid = bill => ({ ...bill, paid: !bill.paid })

export const billList = ({ bill, pay, siteTitle }) => (
  <Paper>
    <Table>
      <TableHead>
        <TableRow>
          <TableCell>paid {siteTitle}</TableCell>
          <TableCell>Biller</TableCell>
          <TableCell numeric>Amount</TableCell>
          <TableCell numeric>Due</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        <TableRow>
          <TableCell>
            <Checkbox checked={bill.paid} onChange={() => pay(handleMarkPaid)} />
          </TableCell>
          <TableCell component="th" scope="row">{bill.name}</TableCell>
          <TableCell numeric>{bill.amount}</TableCell>
          <TableCell numeric>{bill.due}</TableCell>
        </TableRow>
      </TableBody>
    </Table>
  </Paper>
)

export const BillList = compose(
  withState('bill', 'pay', {
    name: 'someone',
    amount: "$1000.00",
    due: 17,
    paid: false,
  })
)(billList)
