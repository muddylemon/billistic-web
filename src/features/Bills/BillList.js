import React from "react"
import "./bill_list.scss"
import { compose, withHandlers, withState } from "recompose"

import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'
import Checkbox from '@material-ui/core/Checkbox'

import { firestoreConnect } from 'react-redux-firebase'
import { connect } from 'react-redux'

const Expenses = ({ expenses, togglePaid }) => expenses.map(expense => (
  <TableRow key={expense.id}>
    <TableCell>
      <Checkbox className="mark_paid" checked={expense.paid} onClick={() => togglePaid(expense.id, !expense.paid)} />
    </TableCell>
    <TableCell component="th" scope="row">{expense.name}</TableCell>
    <TableCell numeric>{expense.amount}</TableCell>
    <TableCell numeric>{expense.due}</TableCell>
  </TableRow>
))

export const billList = ({ expenses, togglePaid }) => (
  <Paper>
    <Table>
      <TableHead>
        <TableRow>
          <TableCell>paid</TableCell>
          <TableCell>Biller</TableCell>
          <TableCell numeric>Amount</TableCell>
          <TableCell numeric>Due</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {!expenses ? (<TableRow><TableCell>Not loaded</TableCell></TableRow>) : (<Expenses expenses={expenses} togglePaid={togglePaid} />)}
      </TableBody>
    </Table>
  </Paper>
)

export const withExpenseHandlers = withHandlers({
  togglePaid: ({ firestore }) => (id, paid = false) => firestore.update(`expenses/${id}`, { paid }),
})

export const withExpenseData = compose(
  firestoreConnect(() => (['expenses'])),
  connect(({ firestore }) => ({
    expenses: firestore.ordered.expenses,
  }))
)

export const BillList = compose(
  withExpenseData,
  withExpenseHandlers
)(billList)
