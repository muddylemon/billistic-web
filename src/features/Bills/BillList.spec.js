import React, { Component } from "react"
import { BillList, withExpenseHandlers } from "./BillList"
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { update } from 'lodash/fp'
import { Provider } from 'react-redux'
import Enzyme, { mount } from 'enzyme'

import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() })

class MockComponent extends Component {
  render() {
    return <div />
  }
}

describe('BillList', () => {
  const initialState = {
    firestore: {
      ordered: {
        expenses: [
          {
            name: 'bank',
            paid: false,
            amount: 20,
            due: 20,
            id: 321,
          },
          {
            name: 'not bank',
            paid: true,
            amount: 20000,
            due: 1,
            id: 123,
          },
        ],
      },
    },
  }
  let store

  beforeEach(() => {
    const mockStore = configureStore()
    store = mockStore(initialState)
  })

  it('should render a new line for each expense', () => {
    const tree = renderer.create(<Provider store={store}><BillList /></Provider>)
    expect(tree).toMatchSnapshot()
  })

  describe('withExpenseHandlers', () => {
    let wrapper, subject

    beforeEach(() => {
      const Enhanced = withExpenseHandlers(MockComponent)
      wrapper = mount(<Enhanced firestore={{ update: jest.fn() }} />)
      subject = wrapper.find(MockComponent)
    })

    it('should togglePaid', () => {
      expect(subject.prop('togglePaid')).toBeTruthy()

      subject.prop('togglePaid')(213, true)

      expect(wrapper.prop('firestore').update).toHaveBeenCalledWith(`expenses/213`, { paid: true })

    })
  })
})
