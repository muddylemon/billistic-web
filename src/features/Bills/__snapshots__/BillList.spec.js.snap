// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillList should render a new line for each expense 1`] = `
<div
  className="MuiPaper-root-1 MuiPaper-elevation2-5 MuiPaper-rounded-2"
>
  <table
    className="MuiTable-root-28"
  >
    <thead
      className="MuiTableHead-root-29"
    >
      <tr
        className="MuiTableRow-root-30 MuiTableRow-head-33"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-head-36"
          scope="col"
        >
          paid
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-head-36"
          scope="col"
        >
          Biller
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-head-36 MuiTableCell-numeric-39"
          scope="col"
        >
          Amount
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-head-36 MuiTableCell-numeric-39"
          scope="col"
        >
          Due
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root-43"
    >
      <tr
        className="MuiTableRow-root-30"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37"
        >
          <span
            className="MuiButtonBase-root-60 MuiIconButton-root-54 MuiPrivateSwitchBase-root-50 MuiCheckbox-root-44 MuiCheckbox-colorSecondary-49 mark_paid"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label-59"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-63"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <input
                checked={false}
                className="MuiPrivateSwitchBase-input-53"
                data-indeterminate={false}
                onChange={[Function]}
                type="checkbox"
              />
            </span>
            <span
              className="MuiTouchRipple-root-72"
            />
          </span>
        </td>
        <th
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37"
          scope="row"
        >
          bank
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37 MuiTableCell-numeric-39"
        >
          20
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37 MuiTableCell-numeric-39"
        >
          20
        </td>
      </tr>
      <tr
        className="MuiTableRow-root-30"
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37"
        >
          <span
            className="MuiButtonBase-root-60 MuiIconButton-root-54 MuiPrivateSwitchBase-root-50 MuiCheckbox-root-44 MuiCheckbox-colorSecondary-49 MuiPrivateSwitchBase-checked-51 MuiCheckbox-checked-45 mark_paid"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label-59"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-63"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
              <input
                checked={true}
                className="MuiPrivateSwitchBase-input-53"
                data-indeterminate={false}
                onChange={[Function]}
                type="checkbox"
              />
            </span>
            <span
              className="MuiTouchRipple-root-72"
            />
          </span>
        </td>
        <th
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37"
          scope="row"
        >
          not bank
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37 MuiTableCell-numeric-39"
        >
          20000
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root-35 MuiTableCell-body-37 MuiTableCell-numeric-39"
        >
          1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
